src/Raven.Server/Documents/ShardedTcpHandlers/ShardedSubscriptionConnection.cs

```diff
@@ -0,0 +1,353 @@
+                        // no batches are currently being sent, we have to ping the client
+                        try
+                        {
+                            await SendHeartBeat("Waited for 3000ms for batch from shard workers");
+                        }
+                        finally
+                        {
+                            _lock.Release();
+                        }
+                    }
+
+                    Dictionary<string, Exception> exceptions = new Dictionary<string, Exception>();
+                    List<string> shardsToReconnect = new List<string>();
+                    var stopping = false;
+                    foreach ((string shard, SubscriptionShardHolder shardHolder) in _shardWorkers)
+                    {
+                        CancellationTokenSource.Token.ThrowIfCancellationRequested();
+
+                        if (shardHolder.PullingTask.IsFaulted == false)
+                            continue;
+
+                        exceptions.Add(shard, shardHolder.PullingTask.Exception);
+
+                        if (stopping)
+                            continue;
+
+                        if (shardHolder.LastErrorDateTime.HasValue == false)
+                        {
+                            shardHolder.LastErrorDateTime = DateTime.Now;
+                            shardsToReconnect.Add(shard);
+                        }
+                        else if (DateTime.Now - shardHolder.LastErrorDateTime.Value < _rejectedReconnectTimeout)
+                        {
+                            shardsToReconnect.Add(shard);
+                        }
+                        else
+                        {
+                            // we are stopping this subscription
+                            stopping = true;
+                        }
+                    }
+
+                    if (stopping)
+                        throw new AggregateException($"Stopping sharded subscription '{_options.SubscriptionName}' with id '{SubscriptionId}' for database '{TcpConnection.ShardedContext.DatabaseName}' because shard {string.Join(", ", exceptions.Keys)} workers failed.", exceptions.Values);
+
+                    // try reconnect faulted workers, this will happen every 3 sec
+                    foreach (var shard in shardsToReconnect)
+                    {
+                        CancellationTokenSource.Token.ThrowIfCancellationRequested();
+
+                        if (_shardWorkers.ContainsKey(shard) == false)
+                            continue;
+
+                        using (var old = _shardWorkers[shard].Worker)
+                        {
+                            _shardWorkers[shard].Worker = new ShardedSubscriptionWorker(_options, shard, _shardWorkers[shard].RequestExecutor, SendConfirmToClientUnderLock, ReportExceptionToAndStopProcessing, WriteBatchToClientAndAck);
+                            var t = _shardWorkers[shard].Worker.RunInternal(CancellationTokenSource.Token);
```

Comment:
you are replacing the worker, but not remembering `LastErrorDateTime` ?

Reply:
I start new subscription worker and assign it to `_shardWorkers[shard].Worker` the error time is not touched its under  `_shardWorkers[shard].LastErrorDateTime`